shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -1799,11 +1857,74 @@ - (void)scrollEvent:(UIPanGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4))
   }
 
+  if (_mouseState.flutter_state_is_added) {
+    packet->SetPointerData(/*index=*/0, pointer_data);
+  } else {
+    flutter::PointerData add_pointer_data = pointer_data;
+    add_pointer_data.signal_kind = flutter::PointerData::SignalKind::kNone;
+    add_pointer_data.change = flutter::PointerData::Change::kAdd;
+    _mouseState.flutter_state_is_added = true;
+    packet->SetPointerData(/*index=*/0, add_pointer_data);
+    packet->SetPointerData(/*index=*/1, pointer_data);
+  }
+
+  [_engine.get() dispatchPointerDataPacket:std::move(packet)];
+}
+
+- (void)panEvent:(UIPanGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4)) {
+  CGPoint translation = [recognizer translationInView:self.view];
+  const CGFloat scale = [UIScreen mainScreen].scale;
+
+  translation.x *= scale;
+  translation.y *= scale;
+
+  flutter::PointerData pointer_data = [self generatePointerDataForMouse];
+  pointer_data.kind = flutter::PointerData::DeviceKind::kTrackpad;
+  pointer_data.device = reinterpret_cast<int64_t>(recognizer);
+  if (recognizer.state == UIGestureRecognizerStateBegan) {
+    pointer_data.change = flutter::PointerData::Change::kPanZoomStart;
+  } else if (recognizer.state == UIGestureRecognizerStateChanged) {
+    pointer_data.change = flutter::PointerData::Change::kPanZoomUpdate;
+    pointer_data.pan_x = translation.x;
+    pointer_data.pan_y = translation.y;
+    pointer_data.pan_delta_x = 0;  // Delta will be generated in pointer_data_packet_converter.cc.
+    pointer_data.pan_delta_y = 0;  // Delta will be generated in pointer_data_packet_converter.cc.
+    pointer_data.scale = 1;
+  } else {
+    pointer_data.change = flutter::PointerData::Change::kPanZoomEnd;
+  }
+
+  auto packet = std::make_unique<flutter::PointerDataPacket>(1);
+  packet->SetPointerData(/*index=*/0, pointer_data);
+
+  [_engine.get() dispatchPointerDataPacket:std::move(packet)];
+}
+
+- (void)pinchEvent:(UIPinchGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4)) {
+  auto packet = std::make_unique<flutter::PointerDataPacket>(1);
+
+  flutter::PointerData pointer_data = [self generatePointerDataForMouse];
+  pointer_data.kind = flutter::PointerData::DeviceKind::kTrackpad;
+  if (recognizer.state == UIGestureRecognizerStateBegan) {
+    pointer_data.change = flutter::PointerData::Change::kPanZoomStart;
+  } else if (recognizer.state == UIGestureRecognizerStateChanged) {
+    pointer_data.change = flutter::PointerData::Change::kPanZoomUpdate;
+    pointer_data.scale = recognizer.scale;
+    pointer_data.rotation = _rotationGestureRecognizer.rotation;
+  } else {
+    pointer_data.change = flutter::PointerData::Change::kPanZoomEnd;
+  }
+
   packet->SetPointerData(/*index=*/0, pointer_data);
 
   [_engine.get() dispatchPointerDataPacket:std::move(packet)];
 }
 
+- (void)rotationEvent:(UIRotationGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4)) {
+  // We need to have an action selector, but the rotation data will instead be queried in
```

Comment:
We need a rotation gesture recognizer to be able to find out the angle of the two fingers on the trackpad. Every time they move, though, it will call the callback of both the pinch and the rotation recognizers. So we can just check both recognizers in one function, the other function is a dummy.

Reply:
Where is the code that we get the rotation angle? 

We specifically set this method to be the selector for the rotation gesture recognizer. I'm still confused how the rotation event is sent to the pinch selector. Am I missing something?

Reply:
`pointer_data.rotation = _rotationGestureRecognizer.rotation;` on line 1913